src/main/scala/org/constellation/util/MajorityStateChooser.scala

```diff
@@ -3,55 +3,100 @@ package org.constellation.util
 
-  val logger = Slf4jLogger.getLogger[F]
+  def compareSnapshotState(major: (Seq[RecentSnapshot], Set[Id]), ownSnapshots: List[RecentSnapshot]) = SnapshotDiff(
+    ownSnapshots.diff(major._1).sortBy(-_.height),
+    major._1.diff(ownSnapshots).toList.sortBy(-_.height),
+    major._2.toList
+  )
+
+  def chooseMajNodeIds(snapshotNodes: SnapshotNodes, nodeSnapshots: List[NodeSnapshots]): Option[Seq[Id]] =
+    nodeSnapshots.filter { case (id, recentSnapshotSeq) => snapshotNodes._2.contains(id) }.sortBy {
+      case (id, recentSnapshotSeq) => recentSnapshotSeq.size
+    }.map { case (id, recentSnapshotSeq) => id } match {
+      case Nil                   => None
+      case majNodIds @ (id :: _) => Some(majNodIds)
+    }
 
-  private final val differenceInSnapshotHeightToReDownloadFromLeader = 10
+  def planReDownload(allProposals: Map[Id, Map[Long, RecentSnapshot]],
+                     peers: Seq[Id],
+                     id: Id): ReDownloadPlan = {
+    val allProposalsNormalized = allProposals.toSeq.flatMap {
+      case (id, proposals: Map[Long, RecentSnapshot]) =>
+        proposals.toSeq.map { case (height, recentSnapshot) => (id, Seq(recentSnapshot)) }
+    }
+    val majority = MajorityStateChooser.chooseMajorWinner(peers, allProposalsNormalized)
+    val groupedProposals = allProposalsNormalized
+      .groupBy { case (id, recentSnapSeq) => id }
+      .map { case (id, tupList) => (id, tupList.flatMap(_._2)) }
+    val snapsThroughMaj = majority.map { case (height, (snap, nodes)) => MajorityStateChooser.getAllSnapsUntilMaj((snap, nodes), groupedProposals.toList) }
+    val majNodeIds =
+      majority.flatMap(maj => MajorityStateChooser.chooseMajNodeIds(maj._2, allProposalsNormalized.toList))
+    val (sortedSnaps, nodeIdsWithSnaps) =
+      (snapsThroughMaj.map(_.sortBy(-_.height)).getOrElse(Seq()), majNodeIds.map(_.toSet).getOrElse(Set()))
+    val diff = MajorityStateChooser.compareSnapshotState((sortedSnaps, nodeIdsWithSnaps), allProposals.getOrElse(id, Map()).values.toList)
+    ReDownloadPlan(id, sortedSnaps, nodeIdsWithSnaps, diff, groupedProposals)
+  }
+
+  def selectMajSnap(totalPeers: Int, allSnapshotNodes: Seq[SnapshotNodes]) = {
+    val numNodesWithSnapAtThisHeight = allSnapshotNodes.flatMap { case (recentSnap, ids) => ids }.length
+    val snapVotes = allSnapshotNodes.map { case (recSnap, ids) => (recSnap, ids.length) }
+    val simpleMaj = snapVotes.find { case (recSnap, count) => count - (totalPeers / 2) >= 0 }.map {
+      case (recSnap, count) => recSnap
+    }
+    if (simpleMaj.isDefined) simpleMaj
+    else if (totalPeers == numNodesWithSnapAtThisHeight) { //todo need actual logic to get majority from a distribution. This could be improved
+      val votedMajority
+        : Option[RecentSnapshot] = snapVotes.sortBy { case (recSnap, count) => (count, recSnap.hash) }.map {
+        case (count, hash) => count
+      }.headOption
+      votedMajority
+    } else None
```

Comment:
Can be simplyfied as `} else None` and `if (simpleMaj.isDefined) simpleMaj` statement branches are redundant (these cases are handled by `Option` out of the box)
```scala
simpleMaj match {
   case None if totalPeers == numNodesWithSnapAtThisHeight => {
      val votedMajority = snapVotes
         .sortBy { case (recSnap, count) => (count, recSnap.hash) }
         .map { case (count, hash) => count }
         .headOption
      votedMajority
   }
   case _ => _
}
```

Reply:
I like it